src/scripts/install_machine.py

```diff
@@ -0,0 +1,330 @@
+
+    if args['startVm']:
+        try:
+            assert_vm_exists(args['connectionName'], args['vmName'])
+
+            xml = virsh(args['connectionName'], "dumpxml", "--inactive", args['vmName'])
+        except subprocess.CalledProcessError:
+            logging.info("The VM got deleted while being installed")
+            logging.info(traceback.format_exc())
+            sys.exit(0)
+
+    xml_last_phase = xml.strip().split(b'\n\n')[-1]
+    # Get last step only - virt-install can output 1 or 2 steps
+    inject_metadata(xml_last_phase.decode())
+
+
+def install_vm(args):
+    # If the VM already exists and we want to run the installer we need to remove it first
+    try:
+        virsh(args['connectionName'], "destroy", args['vmName'])
+    except subprocess.CalledProcessError:
+        pass
+
+    prevXML = virsh(args['connectionName'], "dumpxml", args['vmName'])
+
+    virsh(args['connectionName'], "undefine", args['vmName'], "--managed-save")
+
+    with prepare_virt_install_params(args) as params:
+        try:
+            subprocess.check_output(params)
+        except subprocess.CalledProcessError as e:
+            # If virt-install returned non-zero return code, redefine
+            # the VM so that we get back the metadata which enable the 'Install'
+            # button, so that the user can re-attempt installation
+            with tempfile.NamedTemporaryFile() as file:
+                logging.debug("virt-install failed, redefining to renable the 'Install' button")
+                file.write(prevXML)
+                file.flush()
+                virsh(args['connectionName'], "define", file.name)
+            raise e
+
+        assert_vm_exists(args['connectionName'], args['vmName'])
+
+        xml = virsh(args['connectionName'], "dumpxml", "--inactive", args['vmName'])
+
+        inject_metadata(xml.decode())
+
+
+def inject_metadata(xml):
+    # ET.fromstring() already wants UTF-8 encoded bytes
+    root = ET.fromstring(xml)
+    metadata = root.find('metadata')
+    cockpit_machines_metadata = metadata.find('cockpit_machines')
+    if cockpit_machines_metadata:
+        metadata.remove(cockpit_machines_metadata)
+
+    has_install_phase = "true"
+    # VM does not have a pending install phase (visible Install button in the UI) if:
+    # - The script is called from the 'Install' button
+    # - The script is called from the 'Create' dialog and the VM is started (installer will run)
+    # - The script is called from the 'Import' dialog
+    if args['type'] == 'install' or args['startVm'] or args['sourceType'] == 'disk_image':
+        has_install_phase = "false"
+
+    # Register used namespaces
+    ET.register_namespace("cockpit_machines", "https://github.com/cockpit-project/cockpit-machines")
+    ET.register_namespace("libosinfo", "http://libosinfo.org/xmlns/libvirt/domain/1.0")
+
+    METADATA = f'''
+<cockpit_machines:data xmlns:cockpit_machines="https://github.com/cockpit-project/cockpit-machines"> \
```

Comment:
This still looks weird -- it's declaring the namespace inside a tag which uses it already. I'm not sure if that's legit (but then again libvirt doesn't seem to care)

Reply:
I removed it but then tests failed, so I put it back. I know it's not optimal, but can we make a follow up to clean this up?